### DISCUSSION TITLE
[Feat]: Autoplay - Delay between transitions per slide

### DISCUSSION THREAD

**Discussion by SergkeiM** [davidjerleke/embla-carousel#743] [Ideas] [answered]
### Which variants of Embla Carousel are relevant to this feature request?

- [ ] embla-carousel **(Core)**
- [ ] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [x] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### Feature description

- PR: #1062.

When using `autoplay` there is a posibility to set `delay` between transitions per slide.

I'm trying to change this delay per slide for example (first number is the index of the array of slides | HTMLElement[]):

0 - 7000,
1 - 4000,
2 - 7000,
3 - 10000

I've tried to put `AutoPlay` into a `watchEffect`

```
watchEffect(() => {
        autoplay?.destroy()
        autoplay = null 

        if(props.autoplay){
            autoplay = Autoplay({
                delay: props.delay,
                jump: false,
                playOnInit: true,
                stopOnLastSnap: false,
                stopOnInteraction: false,
                stopOnMouseEnter: false,
                stopOnFocusIn: false
            })
        }

        mergedPlugins.value = [...props.plugins, autoplay].filter(el => !isNull(el))
    })
```
    
 This doesn't work because util function `arePluginsEqual` returns always true as number of plugins didn't change.
 
 I've also tried to `reInit` but this resets the carousel to first slide, I could record the current slide index before that and set it in `reinit`.
 
 @davidjerleke Is there is a way to set a different delay per slide, without all this manipulation?

P.S. Thank you for your work.

**Comment by davidjerleke** [comment]
Hi @SergkeiM,

> When using `autoplay` there is a posibility to set `delay` between transitions per slide.

Thanks for your question. At the time of writing, the Autoplay plugin doesn't support this feature.

Best,
David

**Comment by SergkeiM** [comment]
Hi @davidjerleke 

Any workaround you can suggest?

**Reply by davidjerleke** [reply]
@SergkeiM not anything that comes to mind other than stopping autoplay on every `select` event and run play again manually:

```ts
emblaApi.value.on('select', (emblaApi) => {
  const { autoplay } = emblaApi.value.plugins()
  const currentIndex = emblaApi.selectedScrollSnap() // use this index to determine what delay you want for the next index
  autoplay.stop() // stop autoplay
  autoplay.start() // start this with a timer but be sure to include the current delay in your calculation
})
```

Cumbersome but that's the only way I can think of right now.

Best,
David

**Reply by SergkeiM** [reply]
Thank you @davidjerleke hope you can add this feature in the future.

**Reply by davidjerleke** [reply]
@SergkeiM I’m hoping that too! I have a clear picture of how but I also have a big todo list so let’s see when I get the chance to add the feature!

**Reply by SergkeiM** [reply]
Hi @davidjerleke if you can share how you picture it I can create a PR, to help you out.

Note: Maybe is a good idea (if not already done) share your todolist in milestones, I'm sure people will help you out, the carousel is amazing.

**Reply by sarussss** [reply]
Hi @davidjerleke 
Oh, in the past, I updated the autoplay plugin to support a feature similar to this, can I put it on git for everyone to use?
[Demo](https://stackblitz.com/edit/sarusss-set-custom-autoplay-intervals-per-sli?file=main.js)

**Reply by davidjerleke** [reply]
This has been added to the roadmap:

- #841

**Comment by davidjerleke** [comment, marked as answer]
@SergkeiM and @sarussss this feature is available from [v8.5.0](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.5.0) and up. See the [docs](https://www.embla-carousel.com/plugins/autoplay/#delay).

**Reply by SergkeiM** [reply]
Thank you @davidjerleke long awaited feature, will remove the "hack" after update

**Reply by davidjerleke** [reply]
@SergkeiM please pass a function and return an array of delays for each scroll snap. Example:

```tsx
Autoplay({
  delay: (scrollSnaps) => scrollSnaps.map((scrollSnap, index) => 1000 + (1000 * index))
  // Just for demonstration (you probably don't want this): This will map the delays like this [1000, 2000, 3000, and so on...]
  // So first snap will trigger autoplay after 1000ms, second after 2000ms, third after 3000ms
})
```

**Reply by SergkeiM** [reply]
Hi @davidjerleke 

I'm testing this, out.

This is my config:

```js
const plugins = computed(() => props.autoplay ? [Autoplay({
        delay: (snapList, emblaApi) => slides.value?.map(({ delay }) => (delay || 7) * 1000),
        jump: false,
        playOnInit: true,
        stopOnLastSnap: false,
        stopOnInteraction: false,
        stopOnMouseEnter: false,
        stopOnFocusIn: false
})] : []);
```

And everything works fine except:

Lets say the slide has 7000ms (7s) and you click next on 5th second (`api.scrollNext()`), the next slide will be shown for 2 seconds and slide again.

**Reply by davidjerleke** [reply]
@SergkeiM in your scroll next function you should also call reset, like this:

```tsx 
function myScrollNext() {
  emblaApi.plugins().autoplay.reset()
  emblaApi.scrollNext()
}
```

Feel free to check the [autoplay example](https://www.embla-carousel.com/examples/predefined/#autoplay) out in the docs.

**Reply by SergkeiM** [reply]
Hi @davidjerleke 

Understood, wouldn't be more sense, to reset the autplay behind the scenes on slide change?

**Reply by davidjerleke** [reply]
@SergkeiM the current approach is more flexible because it allows for both behaviors. 

Additionally, I think it’s easier to maintain when the plugin code is separated from the Embla core code. If I start adding many ifs and buts in the core library code that’s related to plugins, the code will be hard to follow and the bundle size of the core library would grow significantly (imagine conditions for all plugins), which would affect many users that don’t use the relevant plugins.

I hope that makes sense.

**Reply by SergkeiM** [reply]
> @SergkeiM the current approach is more flexible because it allows for both behaviors. 
> 
> 
> 
> Additionally, I think it’s easier to maintain when the plugin code is separated from the Embla core code. If I start adding many ifs and buts in the core library code that’s related to plugins, the code will be hard to follow and the bundle size of the core library would grow significantly (imagine conditions for all plugins), which would affect many users that don’t use the relevant plugins.
> 
> 
> 
> I hope that makes sense.

Hi @davidjerleke 

Thank you, makes sense from this POV.
